File: frontend/js/timeline.js
/**
 * timeline.js — CivicGuide AI election timeline feature.
 *
 * Reliability additions in this version:
 *  - Toast notifications for all error states
 *  - Election type validated against known set before API call
 *  - isLoading flag prevents overlapping fetches during tab switching
 *  - Error state rendered in the timeline area (not just console.warn)
 *  - Keyboard accessibility (Space/Enter) for accordion cards
 *  - Null-safe rendering guards on all phase fields
 */

import { apiGet, escapeHtml } from "./api.js";
import { showToast } from "./toast.js";

// ── Colour token → hex map ─────────────────────────────────────────────────
const COLOR_MAP = {
  blue:    "#3b82f6",
  indigo:  "#6366f1",
  violet:  "#8b5cf6",
  cyan:    "#06b6d4",
  amber:   "#f59e0b",
  green:   "#22c55e",
  orange:  "#f97316",
  emerald: "#10b981",
};

// ── Valid election types (mirrors backend SUPPORTED_ELECTION_TYPES) ─────────
const VALID_TYPES = new Set(["lok_sabha", "state_assembly", "rajya_sabha", "local_body"]);

// ── DOM References ─────────────────────────────────────────────────────────
const loadingEl   = document.getElementById("loading-state");
const timelineEl  = document.getElementById("timeline");
const progressBar = document.getElementById("progress-bar");
const metaName    = document.getElementById("meta-name");
const metaFreq    = document.getElementById("meta-freq");
const metaSeats   = document.getElementById("meta-seats");
const metaPhases  = document.getElementById("meta-phases");
const metaAuth    = document.getElementById("meta-auth");

// ── Module State ───────────────────────────────────────────────────────────
let currentType = "lok_sabha";
let openCardId  = null;
let isLoading   = false;  // Prevents overlapping fetches during tab switching


// ══════════════════════════════════════════════════════════════════════════
// DATA LOADING
// ══════════════════════════════════════════════════════════════════════════

/**
 * Fetch the timeline for an election type and render it.
 * Validates the type before calling the API, falls back gracefully on error.
 *
 * @param {string} type - Election type key.
 */
async function loadTimeline(type = "lok_sabha") {
  // Validate the election type — defence against DOM manipulation
  if (!VALID_TYPES.has(type)) {
    showToast(`Unknown election type: "${type}". Defaulting to Lok Sabha.`, "warning");
    type = "lok_sabha";
  }

  // Prevent overlapping requests when the user clicks tabs rapidly
  if (isLoading) return;

  currentType = type;
  isLoading   = true;
  showLoading(true);

  try {
    const data = await apiGet(`/api/timeline?type=${encodeURIComponent(type)}`);

    // Guard: ensure the response has the expected shape
    if (!data.election || !Array.isArray(data.phases)) {
      throw new Error("Unexpected response format from timeline API.");
    }

    renderMeta(data.election);
    renderProgressBar(data.phases);
    renderTimeline(data.phases);

  } catch (err) {
    console.warn("[CivicGuide AI] Timeline API error — using offline fallback:", err.message);

    // Show the offline fallback silently — this is expected when backend is down
    const data = getOfflineFallback();
    renderMeta(data.election);
    renderProgressBar(data.phases);
    renderTimeline(data.phases);

    // Inform the user without being alarmist
    showToast("Showing cached timeline — backend is offline.", "info", 4000);

  } finally {
    isLoading = false;
    showLoading(false);
  }
}


// ══════════════════════════════════════════════════════════════════════════
// RENDER HELPERS
// ══════════════════════════════════════════════════════════════════════════

/** Populate the election meta bar above the timeline. */
function renderMeta(election) {
  metaName.textContent   = election.name           || "—";
  metaFreq.textContent   = election.frequency      || "—";
  metaSeats.textContent  = election.seats          || "—";
  metaPhases.textContent = election.typical_phases || "—";
  metaAuth.textContent   = election.authority      || "—";
}

/**
 * Build the colour-coded progress bar. Each segment scrolls to its phase on click.
 * @param {object[]} phases
 */
function renderProgressBar(phases) {
  if (!phases?.length) return;

  progressBar.innerHTML = phases.map((p) => {
    const color = COLOR_MAP[p.color] || "#6366f1";
    return `<div class="progress-segment"
                 style="background:${color};"
                 title="${escapeHtml(p.name || '')}"
                 data-phase="${escapeHtml(p.id || '')}"
                 role="presentation"></div>`;
  }).join("");

  progressBar.querySelectorAll(".progress-segment").forEach((seg) => {
    seg.addEventListener("click", () => {
      document.getElementById(`phase-${seg.dataset.phase}`)
        ?.scrollIntoView({ behavior: "smooth", block: "center" });
    });
  });
}

/**
 * Render all phase cards and wire interactions.
 * @param {object[]} phases
 */
function renderTimeline(phases) {
  if (!phases?.length) {
    timelineEl.innerHTML = `
      <li style="text-align:center;padding:3rem;color:#64748b;">
        No phases available for this election type.
      </li>`;
    return;
  }

  timelineEl.innerHTML = phases.map((phase, i) => buildPhaseHTML(phase, i)).join("");

  timelineEl.querySelectorAll(".phase-card").forEach((card) => {
    card.addEventListener("click", () => toggleCard(card));
    card.addEventListener("keydown", (e) => {
      if (e.key === " " || e.key === "Enter") {
        e.preventDefault();
        toggleCard(card);
      }
    });
  });

  observePhases();
}

/**
 * Build the HTML string for a single phase list item.
 * All fields are null-safe with || "" fallbacks.
 *
 * @param {object} p - Phase data object.
 * @returns {string}
 */
function buildPhaseHTML(p) {
  const colorClass = `c-${p.color || "blue"}`;

  const activitiesHtml = (p.key_activities || [])
    .map((a) => `
      <li>
        <span class="act-dot ${p.color || "blue"}"></span>
        <span>${escapeHtml(a)}</span>
      </li>`)
    .join("");

  const citizenActionHtml = p.citizen_action
    ? `<div>
         <div class="action-heading">💡 Your Action</div>
         <div class="citizen-action-box ${colorClass}">
           <span class="action-icon" aria-hidden="true">👤</span>
           <span>${escapeHtml(p.citizen_action)}</span>
         </div>
       </div>`
    : "";

  return `
    <li class="phase-item ${colorClass}" id="phase-${escapeHtml(p.id || '')}">

      <div class="phase-dot ${colorClass}" aria-hidden="true">${p.icon || "📌"}</div>

      <div class="phase-card" id="card-${escapeHtml(p.id || '')}"
           role="button" tabindex="0"
           aria-expanded="false"
           aria-controls="body-${escapeHtml(p.id || '')}"
           aria-label="${escapeHtml(p.name || 'Phase')}">

        <div class="card-header ${colorClass}">
          <div class="card-header-left">
            <span class="phase-number">Phase ${p.order || "?"} of 8</span>
            <span class="phase-name">${escapeHtml(p.name || "")}</span>
            <span class="phase-offset ${colorClass}">🕐 ${escapeHtml(p.offset_label || "")}</span>
          </div>
          <div style="display:flex;align-items:flex-start;gap:.75rem;">
            <div class="duration-pill">
              <span class="duration-label">Duration</span>
              <span class="duration-value ${colorClass}">${escapeHtml(p.duration || "")}</span>
            </div>
            <span class="expand-icon">▼</span>
          </div>
        </div>

        <div class="card-body" id="body-${escapeHtml(p.id || '')}" role="region">
          <div class="card-body-inner">
            <p class="phase-description">${escapeHtml(p.description || "")}</p>
            <div>
              <div class="activities-heading">📌 Key Activities</div>
              <ul class="activities-list">${activitiesHtml}</ul>
            </div>
            ${citizenActionHtml}
          </div>
        </div>

      </div>
    </li>`;
}


// ══════════════════════════════════════════════════════════════════════════
// EXPAND / COLLAPSE  (accordion)
// ══════════════════════════════════════════════════════════════════════════

/**
 * Toggle a phase card open/closed. Closes the previous card first.
 * @param {HTMLElement} card
 */
function toggleCard(card) {
  const isOpen = card.classList.contains("open");
  const bodyId = card.id.replace("card-", "body-");
  const body   = document.getElementById(bodyId);

  // Close any previously open card
  if (openCardId && openCardId !== card.id) {
    const prev     = document.getElementById(openCardId);
    const prevBody = document.getElementById(openCardId.replace("card-", "body-"));
    prev?.classList.remove("open");
    prev?.setAttribute("aria-expanded", "false");
    prevBody?.classList.remove("open");
  }

  card.classList.toggle("open", !isOpen);
  card.setAttribute("aria-expanded", String(!isOpen));
  body?.classList.toggle("open", !isOpen);

  openCardId = isOpen ? null : card.id;
}


// ══════════════════════════════════════════════════════════════════════════
// SCROLL REVEAL
// ══════════════════════════════════════════════════════════════════════════

function observePhases() {
  const observer = new IntersectionObserver(
    (entries) => {
      entries.forEach((entry, i) => {
        if (entry.isIntersecting) {
          setTimeout(() => entry.target.classList.add("visible"), i * 80);
          observer.unobserve(entry.target);
        }
      });
    },
    { threshold: 0.1, rootMargin: "0px 0px -40px 0px" }
  );
  timelineEl.querySelectorAll(".phase-item").forEach((el) => observer.observe(el));
}


// ══════════════════════════════════════════════════════════════════════════
// ELECTION TYPE TABS
// ══════════════════════════════════════════════════════════════════════════

document.querySelectorAll(".type-btn").forEach((btn) => {
  btn.addEventListener("click", () => {
    if (isLoading) return;  // Ignore rapid tab clicks during a fetch
    document.querySelectorAll(".type-btn").forEach((b) => b.classList.remove("active"));
    btn.classList.add("active");
    openCardId = null;
    loadTimeline(btn.dataset.type);
  });
});


// ── UI Utilities ───────────────────────────────────────────────────────────

function showLoading(on) {
  loadingEl.style.display  = on ? "flex" : "none";
  timelineEl.style.display = on ? "none" : "ol";
}


// ══════════════════════════════════════════════════════════════════════════
// OFFLINE FALLBACK DATA  (Lok Sabha snapshot)
// ══════════════════════════════════════════════════════════════════════════

function getOfflineFallback() {
  return {
    election: {
      name:           "Lok Sabha (General) Election",
      frequency:      "Every 5 years",
      seats:          543,
      typical_phases: "Up to 7 phases spread across 4–6 weeks",
      authority:      "Election Commission of India",
    },
    phases: [
      { id: "announcement",         order: 1, name: "Election Announcement",      icon: "📢", color: "blue",
        offset_label: "8–10 weeks before Voting Day", duration: "1 day",
        description: "The ECI officially announces the election schedule. The Model Code of Conduct (MCC) comes into effect immediately.",
        key_activities: ["ECI issues election notification.", "MCC becomes effective.", "Government spending restricted.", "Security forces deployed."],
        citizen_action: "Follow ECI's official website for the schedule." },
      { id: "registration_deadline",order: 2, name: "Voter Registration Deadline",icon: "📋", color: "indigo",
        offset_label: "6–7 weeks before Voting Day", duration: "~1 week",
        description: "Last date for citizens to register as new voters or update existing entries.",
        key_activities: ["Submit Form 6 online at voters.eci.gov.in.", "Correct errors via Form 8.", "Electoral Roll published for inspection.", "EROs verify applications."],
        citizen_action: "Check your name at voters.eci.gov.in. Apply using Form 6 before the deadline." },
      { id: "nomination",           order: 3, name: "Nomination Period",           icon: "📝", color: "violet",
        offset_label: "5–6 weeks before Voting Day", duration: "1–2 weeks",
        description: "Candidates file nomination papers with the Returning Officer. Final candidate list is published.",
        key_activities: ["Candidates submit nominations with deposits.", "Returning Officer scrutinises nominations.", "Candidates can withdraw within the window.", "Final candidate list published."],
        citizen_action: "Review candidate affidavits at affidavit.eci.gov.in." },
      { id: "campaign",             order: 4, name: "Campaign Period",             icon: "🗣️", color: "cyan",
        offset_label: "4–5 weeks before Voting Day", duration: "3–5 weeks",
        description: "Parties and candidates campaign through rallies, ads, and door-to-door visits. All activity ceases 48 hrs before polls.",
        key_activities: ["Public rallies and roadshows.", "Print, TV, and digital ads.", "ECI monitors campaign spending.", "Silence period starts 48 hrs before polls."],
        citizen_action: "Read manifestos and attend public meetings. Report violations to the cVIGIL app." },
      { id: "silence_period",       order: 5, name: "Silence Period",              icon: "🤫", color: "amber",
        offset_label: "48 hours before Voting Day", duration: "48 hours",
        description: "All campaign activity must stop 48 hours before polls close. Exit polls are also banned.",
        key_activities: ["All campaign activity ceases.", "No new political advertisements.", "Exit polls are banned.", "Voters check polling booth assignment."],
        citizen_action: "Check your polling booth at voterportal.eci.gov.in and prepare your ID." },
      { id: "voting_day",           order: 6, name: "Voting Day (Poll Day)",       icon: "🗳️", color: "green",
        offset_label: "Day 0", duration: "1 day (7 AM – 6 PM)",
        description: "Eligible registered voters cast ballots using Electronic Voting Machines (EVMs) at polling booths.",
        key_activities: ["Booths open 7 AM, close 6 PM.", "Voters present Voter ID or alternate photo ID.", "Votes cast on EVMs.", "VVPAT slips verify votes."],
        citizen_action: "Carry your Voter ID Card. Vote early! Look for the indelible ink mark after voting." },
      { id: "counting",             order: 7, name: "Vote Counting",               icon: "🔢", color: "orange",
        offset_label: "1–3 days after Voting Day", duration: "1 day",
        description: "EVM votes are counted at designated centres under strict security. Agents of all candidates observe.",
        key_activities: ["Counting begins at ECI-announced time.", "Each round covers one assembly segment.", "Candidate agents monitor tables.", "Live results on ECI's results portal."],
        citizen_action: "Follow live results at results.eci.gov.in." },
      { id: "result_day",           order: 8, name: "Results & New Government",    icon: "🏆", color: "emerald",
        offset_label: "2–5 days after Voting Day", duration: "1–3 weeks",
        description: "ECI declares final results. The winning party is invited to form the government and sworn in.",
        key_activities: ["ECI declares official results.", "President/Governor invites majority party.", "Cabinet formed and sworn in.", "New government begins functioning."],
        citizen_action: "Monitor your elected representative's work via PRS Legislative Research." },
    ],
  };
}


// ══════════════════════════════════════════════════════════════════════════
// INIT
// ══════════════════════════════════════════════════════════════════════════

loadTimeline("lok_sabha");

File: backend/routes/timeline.py
from flask import Blueprint, request, jsonify

timeline_bp = Blueprint("timeline", __name__)

# ── Election Phase Definitions ─────────────────────────────
# Each phase has a fixed order, icon, colour token, and rich metadata.
# Dates are expressed as relative offsets (weeks before Voting Day = 0)
# since actual calendar dates vary per election.

PHASES = [
    {
        "id":          "announcement",
        "order":       1,
        "name":        "Election Announcement",
        "icon":        "📢",
        "color":       "blue",
        "offset_label":"8–10 weeks before Voting Day",
        "duration":    "1 day",
        "description": (
            "The Election Commission of India (ECI) officially announces the election "
            "schedule, including the dates for every phase of voting, the Model Code "
            "of Conduct (MCC) comes into effect immediately."
        ),
        "key_activities": [
            "ECI issues the election notification in the Official Gazette.",
            "Model Code of Conduct (MCC) becomes effective immediately.",
            "Government spending and new policy announcements are restricted.",
            "Security forces are deployed to sensitive areas.",
        ],
        "citizen_action": "Follow news channels and ECI's official website for the schedule.",
    },
    {
        "id":          "registration_deadline",
        "order":       2,
        "name":        "Voter Registration Deadline",
        "icon":        "📋",
        "color":       "indigo",
        "offset_label":"6–7 weeks before Voting Day",
        "duration":    "Approx. 1 week",
        "description": (
            "Last date for citizens to register as new voters (Form 6), "
            "update existing entries (Form 8), or request deletion (Form 7). "
            "After this deadline, no new additions to the Electoral Roll are allowed "
            "for the upcoming election."
        ),
        "key_activities": [
            "New voters: Submit Form 6 online at voters.eci.gov.in or at a BLO office.",
            "Existing voters: Correct name/address errors via Form 8.",
            "Electoral Roll is published and open for public inspection.",
            "Electoral Registration Officers (EROs) verify and approve applications.",
        ],
        "citizen_action": (
            "Check your name at voters.eci.gov.in. If not listed, apply immediately "
            "using Form 6 before the deadline."
        ),
    },
    {
        "id":          "nomination",
        "order":       3,
        "name":        "Nomination Period",
        "icon":        "📝",
        "color":       "violet",
        "offset_label":"5–6 weeks before Voting Day",
        "duration":    "1–2 weeks",
        "description": (
            "Candidates file their nomination papers with the Returning Officer. "
            "Scrutiny of nominations takes place, after which candidates may withdraw. "
            "The final list of candidates is published."
        ),
        "key_activities": [
            "Candidates submit nomination papers with required deposits.",
            "Returning Officer scrutinises all nominations for validity.",
            "Candidates can withdraw nominations within the allowed window.",
            "Final candidate list is published by the ECI.",
        ],
        "citizen_action": (
            "Review the list of candidates in your constituency at "
            "affidavit.eci.gov.in to make an informed vote."
        ),
    },
    {
        "id":          "campaign",
        "order":       4,
        "name":        "Campaign Period",
        "icon":        "🗣️",
        "color":       "cyan",
        "offset_label":"4–5 weeks before Voting Day",
        "duration":    "3–5 weeks",
        "description": (
            "Political parties and candidates actively campaign to win voter support "
            "through rallies, door-to-door visits, advertisements, and social media. "
            "All campaigning must cease 48 hours before voting begins (the 'silence period')."
        ),
        "key_activities": [
            "Candidates hold public rallies, roadshows, and door-to-door campaigns.",
            "Parties run print, TV, and digital media advertisements.",
            "The ECI monitors campaign spending — candidates have a legal limit.",
            "Flying Squads check for cash, liquor, or freebies distribution.",
            "Silence period begins 48 hours before polls open.",
        ],
        "citizen_action": (
            "Attend public meetings, read candidate manifestos, and decide your vote. "
            "Report violations to ECI's cVIGIL app."
        ),
    },
    {
        "id":          "silence_period",
        "order":       5,
        "name":        "Silence Period",
        "icon":        "🤫",
        "color":       "amber",
        "offset_label":"48 hours before Voting Day",
        "duration":    "48 hours",
        "description": (
            "All campaign activity — rallies, advertisements, speeches, and canvassing — "
            "must stop 48 hours before the scheduled close of polls. "
            "Exit polls are also prohibited during this window."
        ),
        "key_activities": [
            "All campaign activity ceases.",
            "No new political advertisements permitted.",
            "Exit polls are banned.",
            "Voters review their final polling booth assignment.",
        ],
        "citizen_action": (
            "Check your polling booth location at voterportal.eci.gov.in and "
            "prepare the documents you'll carry on Voting Day."
        ),
    },
    {
        "id":          "voting_day",
        "order":       6,
        "name":        "Voting Day (Poll Day)",
        "icon":        "🗳️",
        "color":       "green",
        "offset_label":"Day 0",
        "duration":    "1 day (polls open 7 AM – 6 PM typically)",
        "description": (
            "Eligible registered voters cast their ballots at designated polling booths "
            "using Electronic Voting Machines (EVMs). Voters receive an indelible ink "
            "mark on their finger as proof of voting."
        ),
        "key_activities": [
            "Polling booths open at 7:00 AM and close at 6:00 PM (may vary).",
            "Voters present Voter ID (EPIC) or approved alternate photo ID.",
            "Votes are cast on Electronic Voting Machines (EVMs).",
            "Presiding Officer seals EVMs at close of polling.",
            "VVPAT slips allow voters to verify their vote.",
        ],
        "citizen_action": (
            "Carry your Voter ID Card (EPIC) or an approved alternate ID. "
            "Vote early to avoid queues. Look for the indelible ink on your left index finger!"
        ),
    },
    {
        "id":          "counting",
        "order":       7,
        "name":        "Vote Counting",
        "icon":        "🔢",
        "color":       "orange",
        "offset_label":"1–3 days after Voting Day",
        "duration":    "1 day",
        "description": (
            "EVM votes are counted at designated counting centres under strict security. "
            "Agents of all candidates observe the process. "
            "Exit polls and election result predictions are broadcast from this day."
        ),
        "key_activities": [
            "Counting begins at a designated time announced by ECI.",
            "Each round of counting covers one assembly segment.",
            "Candidate agents monitor every counting table.",
            "Trends and results are updated live on ECI's results portal.",
            "Winning candidates receive a certificate of election.",
        ],
        "citizen_action": (
            "Follow live results at results.eci.gov.in or on news channels."
        ),
    },
    {
        "id":          "result_day",
        "order":       8,
        "name":        "Results & New Government",
        "icon":        "🏆",
        "color":       "emerald",
        "offset_label":"2–5 days after Voting Day",
        "duration":    "1–3 weeks (formation of government)",
        "description": (
            "Final results are declared by the ECI. The winning party or coalition "
            "is invited to form the government. The new government is sworn in "
            "within a few weeks of the result."
        ),
        "key_activities": [
            "ECI officially declares final results for all constituencies.",
            "President / Governor invites the majority party/coalition to form the government.",
            "Cabinet is formed and sworn in.",
            "New government presents its agenda and begins functioning.",
        ],
        "citizen_action": (
            "Monitor your elected representative's performance using tools like "
            "the PRS Legislative Research website and your MP's official portal."
        ),
    },
]

ELECTION_TYPES = {
    "lok_sabha": {
        "name":        "Lok Sabha (General) Election",
        "description": "Elections to the lower house of Parliament. Held every 5 years.",
        "frequency":   "Every 5 years",
        "seats":       543,
        "typical_phases": "Up to 7 phases spread across 4–6 weeks",
        "authority":   "Election Commission of India",
    },
    "state_assembly": {
        "name":        "State Legislative Assembly (Vidhan Sabha) Election",
        "description": "Elections to state legislatures. Held every 5 years per state.",
        "frequency":   "Every 5 years (per state)",
        "seats":       "Varies by state",
        "typical_phases": "1–3 phases",
        "authority":   "Election Commission of India",
    },
    "rajya_sabha": {
        "name":        "Rajya Sabha (Upper House) Election",
        "description": "Indirect elections by state MLAs to the upper house of Parliament.",
        "frequency":   "Biennial (1/3 of seats every 2 years)",
        "seats":       "Up to 238 elected seats",
        "typical_phases": "Single phase",
        "authority":   "Election Commission of India",
    },
    "local_body": {
        "name":        "Local Body Election",
        "description": "Elections to municipal corporations, panchayats, and local councils.",
        "frequency":   "Every 5 years",
        "seats":       "Varies by region",
        "typical_phases": "1–2 phases",
        "authority":   "State Election Commission",
    },
}


@timeline_bp.route("/timeline", methods=["GET"])
def get_timeline():
    """
    Return the structured election timeline.

    Query params:
      ?type=lok_sabha | state_assembly | rajya_sabha | local_body
            (default: lok_sabha)
      ?phase=<phase_id>   — return only that specific phase

    Response:
    {
        "election": { ... metadata ... },
        "phases":   [ { ... phase objects ... } ],
        "total_phases": 8,
        "note": "..."
    }
    """
    election_type = request.args.get("type", "lok_sabha").lower().strip()
    phase_id      = request.args.get("phase", "").lower().strip()

    election_meta = ELECTION_TYPES.get(election_type, ELECTION_TYPES["lok_sabha"])

    if phase_id:
        phases = [p for p in PHASES if p["id"] == phase_id]
        if not phases:
            return jsonify({
                "error": f"Phase '{phase_id}' not found.",
                "valid_phases": [p["id"] for p in PHASES]
            }), 404
    else:
        phases = PHASES

    return jsonify({
        "election":     election_meta,
        "election_type": election_type,
        "phases":       phases,
        "total_phases": len(phases),
        "note": (
            "Dates shown are relative to Voting Day. "
            "Actual calendar dates are announced by ECI before each election. "
            "Check eci.gov.in for the official schedule."
        )
    }), 200


@timeline_bp.route("/timeline/phases", methods=["GET"])
def list_phases():
    """Return a summary list of all phases (id + name + icon only)."""
    return jsonify({
        "phases": [
            {"id": p["id"], "name": p["name"], "icon": p["icon"], "order": p["order"]}
            for p in PHASES
        ]
    }), 200
